src/Samples/SubModelSeq/Program.fs

```diff
@@ -146,113 +165,110 @@ module App =
-    { Id = nId
-      Value = Counter.init }
-    |> RoseTree.addChild
+  let hasId id t = t.Data.Id = id
 
   let swapCounters swap nId =
     nId
     |> hasId
     |> List.tryFindIndex
-    |> FuncOption.bindFunc swap
+    |> FuncOption.bind swap
     |> FuncOption.inputIfNone
-    |> RoseTree.mapChildren
+
+  let updateSubtree = function
+    | CounterMsg msg -> msg |> Counter.update |> Identifiable.map |> RoseTree.mapData
+    | AddChild -> createNewLeaf () |> List.cons |> RoseTree.mapChildren
+    | Remove cId -> cId |> hasId >> not |> List.filter |> RoseTree.mapChildren
+    | MoveUp cId -> cId |> swapCounters List.swapWithPrev |> RoseTree.mapChildren
+    | MoveDown cId -> cId |> swapCounters List.swapWithNext |> RoseTree.mapChildren
 
   let update = function
     | ToggleGlobalState -> mapSomeGlobalState not
-    | CounterMsg (nId, msg) -> msg |> Counter.update |> Identifiable.applyIfId nId |> RoseTree.mapData |> mapDummyRoot
-    | AddChild pId -> Guid.NewGuid () |> addChildCounter |> Func.applyIf (hasId pId) |> RoseTree.map |> mapDummyRoot
-    | Remove nId -> nId |> hasId >> not |> List.filter |> RoseTree.mapChildren |> RoseTree.map |> mapDummyRoot
-    | MoveUp nId -> nId |> swapCounters List.swapWithPrev |> RoseTree.map |> mapDummyRoot
-    | MoveDown nId -> nId |> swapCounters List.swapWithNext |> RoseTree.map |> mapDummyRoot
-
-  /// Returns all top-level counters.
-  let topLevelCounters m =
-    m.DummyRoot |> RoseTree.dataOfChildren
-
-  /// Returns all immediate child counters of the specified parent counter ID.
-  let childCountersOf pid m =
-    m.DummyRoot
-    |> RoseTree.flatten
-    |> List.find (fun n -> n.Data.Id = pid)
-    |> RoseTree.dataOfChildren
-
-  /// Returns the parent of the specified child counter ID.
-  let parentOf cid m =
-    m.DummyRoot
-    |> RoseTree.flatten
-    |> List.find (RoseTree.dataOfChildren >> List.map Identifiable.getId >> List.contains cid)
-
-  /// Returns the sibling counters of the specified counter ID.
-  let childrenCountersOfParentOf cid =
-    cid |> parentOf >> RoseTree.dataOfChildren
+    | SubtreeMsg msg -> msg |> RoseTree.update hasId updateSubtree |> mapDummyRoot
 
 
 module Bindings =
 
   open App
 
-  let rec counterTreeBindings () : Binding<Model * Identifiable<Counter>, Msg> list = [
-    "CounterIdText" |> Binding.oneWay(fun (_, c) -> c.Id)
-
-    "CounterValue" |> Binding.oneWay(fun (_, c) -> c.Value.Count)
-    "Increment" |> Binding.cmd(fun (_, c) -> CounterMsg (c.Id, Increment))
-    "Decrement" |> Binding.cmd(fun (_, c) -> CounterMsg (c.Id, Decrement))
+  let canMoveUp (_, (p, c)) =
+    p.Children
+    |> List.tryHead
+    |> Option.map (fun c -> c.Data.Id)
+    |> Option.filter ((<>) c.Data.Id)
+    |> Option.set (MoveUp c.Data.Id)
+
+  let canMoveDown (_, (p, c)) =
+    p.Children
+    |> List.tryLast
+    |> Option.map (fun c -> c.Data.Id)
+    |> Option.filter ((<>) c.Data.Id)
+    |> Option.set (MoveDown c.Data.Id)
+    
+  let adjustMsgToParent msg =
```

Comment:
I'm having trouble understanding why this is needed. Do you have a quick explanation? (I could probably get it if I studied the sample thoroughly, but unfortunately I'm unable to prioritize that for a couple of days at least.)

Reply:
You are very perceptive.  It took me twice as long to think of writing this function than all the other changes in this PR put together.

The signature of this function is of the form `a -> a`, so you could comment out the cases of the pattern match that don't return the input and then run the sample, see what doesn't work, and then think about how you might fix it.  After making that change, clicking any `Remove`, `MoveUp`, and `MoveDown` button becomes a no-op (because some call to `List.mapFirst` doesn't find an element for which the predicate `p` returns `true`).  Maybe you would think of some good solution that I missed.

I also give an explanation below.  The advantage of trying to think of a solution yourself before reading my explanation is that reading my explanation might put you into the same frame of mind that lead to thinking that `adjustMsgToParent` was the "only" solution.

## Explanation

Maybe you remember @ScottHutchinson's example use of the `SubModelSeq` binding to to create a tree.  See  [people/Scott/ShowDialog_subModelSeq](https://github.com/bender2k14/Elmish.WPF/tree/people/Scott/ShowDialog_subModelSeq) if you want a reminder.  In his tree, each node has (at most?) four checkboxes.  Clicking a checkbox toggles the state of that checkbox.  No surprise there.  There are no other ways to modify the tree.  As such, I did not need to use `adjustMsgToParent` in his code.

The possibilities to change the tree in the `SubModelSeq` sample are more complicated.  The simple ones are changes to a counter (`Increment`, `Decrement`, `SetStepSize`, and `Reset`, which are contained in the higher-level message case `CounterMsg`).  They are like toggling the state of a checkbox in @ScottHutchinson's example.  Slightly more complicated (but still "easy") is adding a child, which uses the `AddChild` message case.  Instead of changing the data at a node like the counter changes, this change is "one level up" at the `RoseTree<>`; it adds new data to the list of its children.

The hard cases are when a button click at a node has to change the state of its parent node.  This is the situation for the `Remove`, `MoveUp`, and `MoveDown` message cases.

I think of the function `adjustMsgToParent` as a hack.  If the distance from the root to a certain node is `d`, then clicking a button on that node causes `adjustMsgToParent` to execute `d` times, but only during its first execution does it maybe return something other than the input.

I suggest we keep using `adjustMsgToParent` for now, and I will (eventually) create a new issue to investigate how to get rid of it.

### Solution 1: Change the sample

I recently implemented remove, move up, and move down behavior for a list in my application at work.  Instead of having buttons on each item in the list, there is one button for each operation above the list, a selected item, and the action of the buttons are for the selected item.

We could make this change to the sample.  This removes the need for `adjustMsgToParent` but adds a new one.

It would be difficult for a `SubModelSelectedItem` binding to construct the correct `RoseTreeMsg<,>` value.  Instead, it would be convenient for to change `SubModelSelectedItemData<...>.Set` from...
```F#
'id voption -> 'model -> 'msg
```

...to...
```F#
'id voption -> 'model -> ('id * 'bindingMsg -> 'msg) -> 'msg
```

...where the new argument is `SubModelSeqData<...>.ToMsg`.

This non-breaking change API is trivial to make.

We could also make this trivial change, add the remove / move up / move down buttons for the selected item, and also keep the all the existing remove / move up / move down buttons.

### Solution 2;

The current approach to construct the final `RoseTreeMsg<,>` value has the `CounterMsg` and `AddChild` cases "work by default" and makes the `Remove`, `MoveUp`, and `MoveDown` cases "do extra work".  I wanted to reverse this, but I don't see how this is currently possible.

I think adding a `Binding.subModelSeq` overload would make this possible.  The most expressive overload of this function has...
```F#
toMsg: 'id * 'bindingMsg -> 'msg
```

Adding an overload with...
```F#
toMsg: 'bindingModel -> 'id * 'bindingMsg -> 'msg
```

...should be enough.

(Interesting, I keep finding reasons to pass more data into these `'msg` creating functions.  First `'model` when creating `mapMsgWithModel` and now `'bindingModel` in this case.)

Reply:
> The signature of this function is of the form `a -> a`, so you could comment out the cases of the pattern match that don't return the input and then run the sample, see what doesn't work, and then think about how you might fix it.

I can kind of do this for you with screenshots.

## With `adjustMsgToParent`
### Before clicking remove
![With_adjustMsgToParent_Before](https://user-images.githubusercontent.com/34664007/90635206-7f27c400-e1ee-11ea-9433-f13cc6d7388e.png)

### After clicking remove
![With_adjustMsgToParent_After](https://user-images.githubusercontent.com/34664007/90635225-85b63b80-e1ee-11ea-95e2-4f97d3f05314.png)

The item was removed.  This is the expected behavior.

## Without `adjustMsgToParent`
### Before clicking remove
![Without_adjustMsgToParent_Before](https://user-images.githubusercontent.com/34664007/90635264-95358480-e1ee-11ea-99df-b2ee23ad4ac2.png)

### After clicking remove
![Without_adjustMsgToParent_After](https://user-images.githubusercontent.com/34664007/90635284-9a92cf00-e1ee-11ea-9f85-a5a10e4e47c7.png)

The item was not removed.  The button click was a no-op.  This is _not_ the expected behavior.  The message is not correct.  The no-op occured because `mapFirst` returned the same list it was given when (via the code `match back with [] -> input | ...`).